### PR TITLE
HAWQ-1603.add new hook API

### PR DESCRIPTION
**Pull request by zhangjackey**
add two hook API.

### CODE REVIEW

File: contrib/vexecutor/execVQual.c [hunk 1]
```diff
@@ -123,3 +123,431 @@ VirtualNodeProc(ScanState* state,TupleTableSlot *slot){
...
+
+	/* isNull is a single value, it can not be used when data is vectorized */
+	*isNull = false;
+
+	/* Fetch the value from the slot */
+	tb = (TupleBatch )slot->PRIVATE_tb;
+
+	Assert(NULL != tb);
+
+	return PointerGetDatum(tb->datagroup[attnum]);
```
Comment (wengyanqing) [inline contrib/vexecutor/execVQual.c]: attnum's value starts with 1, it should be attnum -1 as the index of tb->datagroup ?

File: contrib/vexecutor/execVQual.c [hunk 2]
```diff
@@ -123,3 +123,431 @@ VirtualNodeProc(ScanState* state,TupleTableSlot *slot){
...
+				bstate->args = (List *)
+					ExecInitExpr((Expr *) boolexpr->args, parent);
+				state = (ExprState *) bstate;
+			}
+			break;
+		/*TODO: More and more expressions should be vectorized */
+		default:
+			break;
+	}
+
```
Comment (wengyanqing) [inline contrib/vexecutor/execVQual.c]: Does it need "state->expr = node;" here?

File: contrib/vexecutor/execVQual.c [hunk 3]
```diff
@@ -123,3 +123,431 @@ VirtualNodeProc(ScanState* state,TupleTableSlot *slot){
...
+						break;
+				}
+				bstate->args = (List *)
+					ExecInitExpr((Expr *) boolexpr->args, parent);
+				state = (ExprState *) bstate;
+			}
+			break;
+		/*TODO: More and more expressions should be vectorized */
+		default:
+			break;
```
Comment (wengyanqing) [inline contrib/vexecutor/execVQual.c]: It's better to think about how to fallback to original executor when we find any unsupported behavior to make sure the query could have right result. There are so many places that just error out or assert(false) which is not friendly.

File: contrib/vexecutor/vexecutor.c [hunk 4]
```diff
@@ -163,24 +165,35 @@ static PlanState* VExecInitNode(PlanState *node,EState *eState,int eflags,Memory
 			case T_AppendOnlyScan:
 			case T_ParquetScan:
 			case T_TableScanState:
-				START_MEMORY_ACCOUNT(curMemoryAccount);
+				//START_MEMORY_ACCOUNT(curMemoryAccount);
```
Comment (wengyanqing) [inline contrib/vexecutor/vexecutor.c]: Any reason that comment this code line ?

File: contrib/vexecutor/execVQual.c [hunk 5]
```diff
@@ -123,3 +123,431 @@ VirtualNodeProc(ScanState* state,TupleTableSlot *slot){
...
+
+		default:				/* get the tuple from the relation being
+								 * scanned */
+			slot = econtext->ecxt_scantuple;
+			break;
+	}
+
+	/* isNull is a single value, it can not be used when data is vectorized */
+	*isNull = false;
+
```
Comment (wengyanqing) [inline contrib/vexecutor/execVQual.c]: It's better to check or assert to make sure slot is not null here.

File: contrib/vexecutor/execVQual.c (hunk 2 above)
Comment (zhangjackey) [reply contrib/vexecutor/execVQual.c]: yes, add in next commit.

File: contrib/vexecutor/execVQual.c (hunk 5 above)
Comment (zhangjackey) [reply contrib/vexecutor/execVQual.c]: add Assert in next commit.

File: contrib/vexecutor/execVQual.c (hunk 1 above)
Comment (zhangjackey) [reply contrib/vexecutor/execVQual.c]: yes, change it to attnum - 1.

**Pull request closed without merging**